src/violation.rs

```diff
@@ -0,0 +1,60 @@
+use std::fmt::Debug;
```

Comment:
Given the discussion in #1546, my understanding is that everywhere we used `Violation` in this PR should be called `Rule`, and that the `Check` struct should be renamed to violation. Is that right? Am I misunderstanding?

Reply:
Yes I think you are misunderstanding.

The `self` in the trait refers to a specific violation, so I think it makes sense to also name the trait `Violation`.
This also means we cannot name the `Check` struct "Violation" as well since that would lead to [E0252](https://doc.rust-lang.org/stable/error_codes/E0252.html).

I agree that `Check` should be renamed ... I am starting to think that `Diag` (short for Diagnostic) would make sense.

Reply:
Ah, right, ok. This makes sense. Does "rule" exist in the codebase at all then? Would a "rule" be the function that takes in the source code data (AST nodes, tokens, whatever) and generates diagnostics?

Reply:
Yes I'd consider the functions that create the diagnostics to be the definitions of the rules.